Project setup
=============

Group
-----
Amanda Charbonneau, Tracy Teal, Laura Williams, Juan Ugalde, ...

Learning Objectives:
-------------------

#### What's the motiviation for this lesson?
* Equip learners with knowledge and skills to organize and access data so that they and their colleagues can use it
* Keeping you data accessible to you. Sequencing centers don't keep the data (for example), and you might need to redo analysis.
* Keep track of your analysis. In two months, you won't remember what you did with your data.
* Sharing with collaborators is less fraught with anxiety that you're giving them the right version or the clean version
* When you go to publish your paper, you have your data ready to put in the data repository
* Journals and databases are requiring this information
* Makes you feel happy 

#### What mindset change should the learner have? (ie, awareness of certain techniques)

* Time spent organizing data is time that should be valued https://xkcd.com/1205/
* Value data organization   
* Good data organization saves you time
* Preventative data care
* The computer is a tool that you're doing experiments with - you can think about the data analysis as a fundamental part of the experiment and should extend the same good practices 

#### What skills should the learner leave the room w/?
* How to set up their computational environmental for the types of experiments you're going to do
* metadata organization in spreadsheets. For example, keep a master list of genomic data
* Awareness of metadata standards (where is that info). What minimal information do you need
* Management of raw data files, including:
  * Good naming schemes with unique names
  * No spaces or non-normal characters
  * Backed up
  * Accessible to lab mates
  * Don't mess with raw data files
* Knowing levels of data backup
* Knowing places where you can put your data
* management of analysis files
* Documenting your steps / data exploration (this could be incorporated in to data wrangling)
  * What you did and why


#### What would be an exercise:

* Paired exercise - each of you do an analysis, then swap and see if you can reproduce it.
* Paired exercise - work with a partner to develop the beginnings of a scheme for organizing and storing your datasets (each attendee probably has a general project plan, a partner could be a good sounding board for developing an organizational structure for data)
* Organize the data related with the test data set *E. coli* from [Patric] (http://patricbrc.org/portal/portal/patric/Home). This can be used as a test example. This can be done in collaboration with a lab partner, where you can try to reproduce, understand the data generated.


#### What are the pre-reqs
* Knowing what kind of data you'll be dealing with
* Knowing what kinds of things you need to keep track of for your experiments, what do people in your field care about
vocabulary:
  * what are different levels of backups
  * metadata
  * what do we mean by raw data

#### Other comments

##### Recommendations (good, better, best)
* meaningful file names
* README files
* automated backups
* pass 'leaving academia forever' test


provenance

opportunities for data storage/management through other resources
- github https://github.com/blog/1840-improving-github-for-science
- dat
- osf.io
- dryad
- bitbucket (offers free private repos for academic users)

#### Assessment Questions

7-10 key things attitudes/dispositions vs declarative knowledge vs skill  
"I feel" vs "I  know this about X " vs "I know how to do X" vs   
Please remember (Where possible): Be measurable, specific, precise. 

#####Attitudes
- I feel that having #reproducible data is worth the time collating the metadata  
Rate your feelings from 1 to 5(most true): 
 - If necessary, it will be easy to re-process my sequencing data as long as I know what programs I used, and in what order
 - If two people analyze sequencing data using the same set of tools they will always get the same answer
 - I need to write down which programs I use to process my sequencing data, and in what order
 - My computer automatically records the steps of my data processing
 - I can predict what would be in my colleagues "e-coli assembly" folder, without looking
 - A readme file should be written only after all the work is finished, so it will be well-written and complete

- I think that documenting my digital data processing is as important as documenting my laboratory data processing  
what I did and why
 - How long do you typically spend documenting your computational processing (sequencing data cleaning at several cutoffs)?  Your laboratory processing (PCR experiment with many temps/salt concentrations)?
 - I think of my digital data processing as hypothesis testing and a series of experiments
- How many times to you expect to process your sequencing data before you'll believe it's accurate? How many times do you typically amplify a gene sequence before you'll believe it's accurate?

#####Skills
I can identify shortcomings in other datasets
give the Patric 0104H4 database and ask them to organize the metadata for their own use. Later have them revisit their results and pick out some data, or pick out data from a neighbor to demo how easy it is to get lost in your own poorly organized filing system (see below for detailed example)
Patric 0104H4 Dataset: http://patricbrc.org/portal/portal/patric/GenomeList?cType=taxon&cId=1038927&displayMode=&dataSource=PATRIC&pk=#aP0=1&aP1=1&aT=0&key=1349703250&cwG=false&tId=1038927&gName=&kW=
I can transform badly structured data into well structured data
Have the students restructure their first try at organizing the Patric 0104H4 dataset
I can set up a cohesive and well-documented data collection system and consistently apply it to projects
Have students add a hypothetical new isolate to the Patric O104:H4 database
From an already populated file structure with data,




